## Association Rule Mining in Retail Store

### Problem Statement:
 * What are the items that may be frequently purchased together?

### Objective:
* To learn how Apriori Algorithm and Association Rules works.
* To learn how Combination and Permutation helps to find Support and Confidence of itemsets respectively.
* To find frequent itemsets with high confidence and lift, keeping both item together will help to increase sales.


### Introduction
* Association rule mining is one of an important technique of data mining for knowledge discovery.
* The knowledge of the correlation between the items in the data transaction can use association rule mining.
* Retail store analysis is one of an application area of association rule mining technique.
* The possible percentage of the correlation of combined items gives the new knowledge. Therefore, it is a very helpful for determiner to take the decisions

### Analysis

In [1]:
# Importing Required Library

import pandas as pd
import numpy as np

In [2]:
# Reading Excel file 
# need to install 1.2.0 version xlrd package to read excel files: pip install xlrd==1.2.0
bread = pd.read_excel('raw_bread.xlsx')

In [3]:
## Here we have transaction data, which include column, Date,Time,Transaction,Item
## we should remove duplicate transaction, it shows quantity of item in same transaction,
## it is not needed in appriori aglo as we only care about different item in particular transaction
bread

Date,Time,Transaction,Item
0              2016-10-30,09:58:11,1,Bread
1       2016-10-30,10:05:34,2,Scandinavian
2       2016-10-30,10:05:34,2,Scandinavian
3      2016-10-30,10:07:57,3,Hot chocolate
4                2016-10-30,10:07:57,3,Jam
...                                    ...
21288      2017-04-09,14:32:58,9682,Coffee
21289         2017-04-09,14:32:58,9682,Tea
21290      2017-04-09,14:57:06,9683,Coffee
21291      2017-04-09,14:57:06,9683,Pastry
21292   2017-04-09,15:04:24,9684,Smoothies

[21293 rows x 1 columns]

In [4]:
## dropping Duplicate Transaction
bread = bread.drop_duplicates()

In [5]:
## we need to split transaction data into Dataframe/tabular structure as follow
new = bread['Date,Time,Transaction,Item'].str.split(',', n = 3, expand = True)

In [6]:
import warnings
warnings.filterwarnings('ignore')

In [7]:
## assigning column to data frame "bread"
bread['Date'] = new[0]
bread['Time'] = new[1]
bread['Transaction'] = new[2]
bread['Item'] = new[3]

In [8]:
# in this dataframe we only need column Trasaction and Item, rest is not needed in association mining rule
bread[['Date', 'Time', 'Transaction', 'Item']].head(10)
                                                    

Date      Time Transaction           Item
0   2016-10-30  09:58:11           1          Bread
1   2016-10-30  10:05:34           2   Scandinavian
3   2016-10-30  10:07:57           3  Hot chocolate
4   2016-10-30  10:07:57           3            Jam
5   2016-10-30  10:07:57           3        Cookies
6   2016-10-30  10:08:41           4         Muffin
7   2016-10-30  10:13:03           5         Coffee
8   2016-10-30  10:13:03           5         Pastry
9   2016-10-30  10:13:03           5          Bread
10  2016-10-30  10:16:55           6      Medialuna

In [9]:
# we need to convert cloumn transacton & item into Crosstab or we can say Binary Matrix as follow
transaction = pd.crosstab(index= bread['Transaction'], columns= bread['Item'])
transaction

Item         Adjustment  Afternoon with the baker  Alfajores  Argentina Night  \
Transaction                                                                     
1                     0                         0          0                0   
10                    0                         0          0                0   
100                   0                         0          0                0   
1000                  0                         0          0                0   
1001                  0                         0          0                0   
...                 ...                       ...        ...              ...   
995                   0                         0          0                0   
996                   0                         0          0                0   
997                   0                         0          0                0   
998                   0                         0          0                0   
999                   0                         0          1                0   

Item         Art Tray  Bacon  Baguette  Bakewell  Bare Popcorn  Basket  ...  \
Transaction                                                             ...   
1                   0      0         0         0             0       0  ...   
10                  0      0         0         0             0       0  ...   
100                 0      0         0         0             0       0  ...   
1000                0      0         0         0             0       0  ...   
1001                0      0         0         0             0       0  ...   
...               ...    ...       ...       ...           ...     ...  ...   
995                 0      0         0         0             0       0  ...   
996                 0      0         0         0             0       0  ...   
997                 0      0         0         0             0       0  ...   
998                 0      0         0         0             0       0  ...   
999                 0      0         0         0             0       0  ...   

Item         The BART  The Nomad  Tiffin  Toast  Truffles  Tshirt  \
Transaction                                                         
1                   0          0       0      0         0       0   
10                  0          0       0      0         0       0   
100                 0          0       0      0         0       0   
1000                0          0       0      0         0       0   
1001                0          0       0      0         1       0   
...               ...        ...     ...    ...       ...     ...   
995                 0          0       0      0         0       0   
996                 0          0       0      0         1       0   
997                 0          0       0      0         0       0   
998                 0          0       0      0         0       0   
999                 0          0       0      0         0       0   

Item         Valentine's card  Vegan Feast  Vegan mincepie  Victorian Sponge  
Transaction                                                                   
1                           0            0               0                 0  
10                          0            0               0                 0  
100                         0            0               0                 0  
1000                        0            0               0                 0  
1001                        0            0               0                 0  
...                       ...          ...             ...               ...  
995                         0            0               0                 0  
996                         0            0               0                 0  
997                         0            0               0                 0  
998                         0            0               0                 0  
999                         0            0               0              

In [10]:
## Just writing csv file to check result
## we have one unwanted column named "NONE", we should remove it as follow and proceed further
transaction.to_csv('tab.csv')

In [11]:
## removing unwanted col "NONE"
transaction = transaction.drop(['NONE'], axis = 1)

In [12]:
transaction

Item         Adjustment  Afternoon with the baker  Alfajores  Argentina Night  \
Transaction                                                                     
1                     0                         0          0                0   
10                    0                         0          0                0   
100                   0                         0          0                0   
1000                  0                         0          0                0   
1001                  0                         0          0                0   
...                 ...                       ...        ...              ...   
995                   0                         0          0                0   
996                   0                         0          0                0   
997                   0                         0          0                0   
998                   0                         0          0                0   
999                   0                         0          1                0   

Item         Art Tray  Bacon  Baguette  Bakewell  Bare Popcorn  Basket  ...  \
Transaction                                                             ...   
1                   0      0         0         0             0       0  ...   
10                  0      0         0         0             0       0  ...   
100                 0      0         0         0             0       0  ...   
1000                0      0         0         0             0       0  ...   
1001                0      0         0         0             0       0  ...   
...               ...    ...       ...       ...           ...     ...  ...   
995                 0      0         0         0             0       0  ...   
996                 0      0         0         0             0       0  ...   
997                 0      0         0         0             0       0  ...   
998                 0      0         0         0             0       0  ...   
999                 0      0         0         0             0       0  ...   

Item         The BART  The Nomad  Tiffin  Toast  Truffles  Tshirt  \
Transaction                                                         
1                   0          0       0      0         0       0   
10                  0          0       0      0         0       0   
100                 0          0       0      0         0       0   
1000                0          0       0      0         0       0   
1001                0          0       0      0         1       0   
...               ...        ...     ...    ...       ...     ...   
995                 0          0       0      0         0       0   
996                 0          0       0      0         1       0   
997                 0          0       0      0         0       0   
998                 0          0       0      0         0       0   
999                 0          0       0      0         0       0   

Item         Valentine's card  Vegan Feast  Vegan mincepie  Victorian Sponge  
Transaction                                                                   
1                           0            0               0                 0  
10                          0            0               0                 0  
100                         0            0               0                 0  
1000                        0            0               0                 0  
1001                        0            0               0                 0  
...                       ...          ...             ...               ...  
995                         0            0               0                 0  
996                         0            0               0                 0  
997                         0            0               0                 0  
998                         0            0               0                 0  
999                         0            0               0              

### Creating APRIORI function to generate frequent itesets based on minimum threshold support = 0.02

In [13]:
def APRIORI(data, min_support = 0.04,  max_length = 4):
    # Collecting Required Library
    import numpy as np
    import pandas as pd
    from itertools import combinations
    
    support = {} 
    L = list(data.columns)
    total_items=len(data)
    # write your code here !!
    k=1
    while k<=max_length:
        can=list(combinations(L,k))
        L=[]
        for l in range(len(can)):
            i=0
            df1=data.loc[data[can[l][i]]==1]
            df2=[]
            df3=df1
            while i<k-1:
                df2=df1.loc[data[can[l][i+1]]==1]
                i=i+1
                df1=df2
            if k!=1:
                df3=df2
            sup=(len(df3))/total_items
            if sup >= min_support:
                support[can[l]]=sup
                for g in can[l]:
                    L.append(g)
        L=list(set(L))  
        if L==[]:
            break
        k=k+1
    result = pd.DataFrame(list(support.items()), columns = ["Items", "Support"])
    return result

In [17]:
import numpy as np
import pandas as pd
from itertools import combinations

support = {} 
L = list(transaction.columns)
total_items=len(transaction)
# write your code here !!
k=1
can=list(combinations(L,k))
print(can)
L=[]
for l in range(len(can)):
    i=0
    df1=transaction.loc[transaction[can[l][i]]==1]
    df2=[]
    df3=df1
#     print(df1)
    while i<k-1:
        df2=df1.loc[transaction[can[l][i+1]]==1]
        i=i+1
        df1=df2
    if k!=1:
        df3=df2
    sup=(len(df3))/total_items
    if sup >= 0.04:
        support[can[l]]=sup
        for g in can[l]:
            L.append(g)
L=list(set(L))  

[('Adjustment',), ('Afternoon with the baker',), ('Alfajores',), ('Argentina Night',), ('Art Tray',), ('Bacon',), ('Baguette',), ('Bakewell',), ('Bare Popcorn',), ('Basket',), ('Bowl Nic Pitt',), ('Bread',), ('Bread Pudding',), ('Brioche and salami',), ('Brownie',), ('Cake',), ('Caramel bites',), ('Cherry me Dried fruit',), ('Chicken Stew',), ('Chicken sand',), ('Chimichurri Oil',), ('Chocolates',), ('Christmas common',), ('Coffee',), ('Coffee granules ',), ('Coke',), ('Cookies',), ('Crepes',), ('Crisps',), ('Drinking chocolate spoons ',), ('Duck egg',), ('Dulce de Leche',), ('Eggs',), ("Ella's Kitchen Pouches",), ('Empanadas',), ('Extra Salami or Feta',), ('Fairy Doors',), ('Farm House',), ('Focaccia',), ('Frittata',), ('Fudge',), ('Gift voucher',), ('Gingerbread syrup',), ('Granola',), ('Hack the stack',), ('Half slice Monster ',), ('Hearty & Seasonal',), ('Honey',), ('Hot chocolate',), ('Jam',), ('Jammie Dodgers',), ('Juice',), ('Keeping It Local',), ('Kids biscuit',), ('Lemon and c

Item         Adjustment  Afternoon with the baker  Alfajores  Argentina Night  \
Transaction                                                                     
1026                  0                         0          0                0   
1036                  0                         0          0                0   
1087                  0                         0          0                0   
1090                  0                         0          0                0   
1093                  0                         0          0                0   
...                 ...                       ...        ...              ...   
9654                  0                         0          0                0   
972                   0                         0          0                0   
983                   0                         0          1                0   
993                   0                         0          0                0   
996                   0     

[13 rows x 94 columns]
Item         Adjustment  Afternoon with the baker  Alfajores  Argentina Night  \
Transaction                                                                     
6000                  0                         0          0                0   
6608                  0                         0          0                0   
7363                  0                         0          0                0   
7563                  0                         0          0                0   
7880                  0                         0          0                0   
7981                  0                         0          0                1   
8007                  0                         0          0                0   
8161                  0                         0          0                0   
8888                  0                         0          0                0   
9637                  0                         0          0                0   
9638 

In [17]:
## finding frequent itemset with min support = 4%
my_freq_itemset = APRIORI(transaction, 0.04, 3)
my_freq_itemset.sort_values(by = 'Support', ascending = False)

Items   Support
2          (Coffee,)  0.475081
0           (Bread,)  0.324940
8             (Tea,)  0.141643
1            (Cake,)  0.103137
12   (Coffee, Bread)  0.089393
6          (Pastry,)  0.085510
7        (Sandwich,)  0.071346
5       (Medialuna,)  0.061379
4   (Hot chocolate,)  0.057916
9     (Cake, Coffee)  0.054349
3         (Cookies,)  0.054034
11     (Tea, Coffee)  0.049523
10  (Pastry, Coffee)  0.047214

### Creating ASSOCIATION_RULE function to generate rules based on minimun threshold confidence.

In [161]:
def ASSOCIATION_RULE(df, min_confidence_threshold = 0.5):
    import pandas as pd
    from itertools import permutations
    
    support = pd.Series(df.Support.values, index=df.Items).to_dict()
    data = []
    L = df.Items.values
    
    # write your code here !!
    
    L_p=list(permutations(L,2))
    for i in L_p:
        if len(i[0])<2 and len(i[1])<2:
            tup=(i[0][0],i[1][0])
            tup_p=list(permutations(tup))
        elif len(i[0])<2 and len(i[1])>=2:
            tup=(i[0][0],i[1][0],i[1][1])
            tup_p=list(permutations(tup))
        elif len(i[0])>=2 and len(i[1])<2:
            tup=(i[0][0],i[0][1],i[1][0])
            tup_p=list(permutations(tup))
        elif len(i[0])>=2 and len(i[1])>=2:
            tup=(i[0][0],i[0][1],i[1][0],i[1][1])
            tup_p=list(permutations(tup))
        for k in tup_p:
            if support.get(k) != None:
                down=support[i[0]]
                con=support.get(k)/down
                if con>min_confidence_threshold:
                    if con>0.5:
                        supa=support[i[0]]
                        supc=support[i[1]]
                        supac=support.get(k)
                        lift=con/supc
                        lev=supac-supa*supc
                        convec=(1-supc)/(1-con)
                        df=[i[0],i[1],supa,supc,supac,con,lift,lev,convec]
                        data.append(df)
#                         print(df)
            else:
                continue
    
    result = pd.DataFrame(data, columns = ["antecedents", "consequents", "antecedent support", "consequent support",
                                        "support", "confidence", "Lift", "Leverage", "Convection"])
    return result

In [162]:
## Rule with minimun confidence = 50%
my_rule = ASSOCIATION_RULE(my_freq_itemset, 0.5)
my_rule
# the column names in the following table is the same as those generated by mlxtend
# you can refer to http://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/ 
# to find the formulas of lift, leverage, convection

antecedents consequents  antecedent support  consequent support   support  \
0     (Cake,)   (Coffee,)            0.103137            0.475081  0.054349   
1   (Pastry,)   (Coffee,)            0.085510            0.475081  0.047214   

   confidence      Lift  Leverage  Convection  
0    0.526958  1.109196   0.00535    1.109667  
1    0.552147  1.162216   0.00659    1.172079

### Finally sorting results by Lift to get highly associated itemsets.

In [163]:
my_rule.sort_values(by='Lift', ascending= False).head(10)

antecedents consequents  antecedent support  consequent support   support  \
1   (Pastry,)   (Coffee,)            0.085510            0.475081  0.047214   
0     (Cake,)   (Coffee,)            0.103137            0.475081  0.054349   

   confidence      Lift  Leverage  Convection  
1    0.552147  1.162216   0.00659    1.172079  
0    0.526958  1.109196   0.00535    1.109667

## Cross Verifying results with  apriori and association rule from mlxtend

In [166]:
# Loading standard package
import mlxtend
from mlxtend.frequent_patterns import apriori, association_rules

In [167]:
## finding frequent itemset with min support = 4%
frequent_itemset = apriori(df = transaction, min_support= 0.04, use_colnames= True)
frequent_itemset.sort_values(by = 'support', ascending = False)

support          itemsets
2   0.475081          (Coffee)
0   0.324940           (Bread)
8   0.141643             (Tea)
1   0.103137            (Cake)
9   0.089393   (Coffee, Bread)
6   0.085510          (Pastry)
7   0.071346        (Sandwich)
5   0.061379       (Medialuna)
4   0.057916   (Hot chocolate)
10  0.054349    (Coffee, Cake)
3   0.054034         (Cookies)
12  0.049523     (Coffee, Tea)
11  0.047214  (Coffee, Pastry)

### Createing associate rule such that item brought with conditional probability(Confidence) more than 50% with corresponding item

In [168]:
## Rule with minimun confidence = 50%
Rules = association_rules(frequent_itemset, min_threshold= 0.5)
Rules

antecedents consequents  antecedent support  consequent support   support  \
0      (Cake)    (Coffee)            0.103137            0.475081  0.054349   
1    (Pastry)    (Coffee)            0.085510            0.475081  0.047214   

   confidence      lift  leverage  conviction  
0    0.526958  1.109196   0.00535    1.109667  
1    0.552147  1.162216   0.00659    1.172079

In [169]:
## Finally sorting results by Lift to get highly associated itemsets.
Rules.sort_values(by='lift', ascending= False).head(10)

antecedents consequents  antecedent support  consequent support   support  \
1    (Pastry)    (Coffee)            0.085510            0.475081  0.047214   
0      (Cake)    (Coffee)            0.103137            0.475081  0.054349   

   confidence      lift  leverage  conviction  
1    0.552147  1.162216   0.00659    1.172079  
0    0.526958  1.109196   0.00535    1.109667

### Conclusion
 * Results from developed function(APRIORI, ASSOCIATION_RULE) has matched with builts packages.
 * it is observed that "Toast" & "Coffee" are highly associated with lift 1.48.
 * Coffee has been brought most frequently with 47.5% of all the transactions